src/analyzer.ts

```diff
@@ -1180,6 +1183,115 @@ export class Analyzer {
     }
   }
 
+  private _processEnv(
+    placeholder: UnvalidatedConfig,
+    packageJson: PackageJson,
+    syntaxInfo: ScriptSyntaxInfo,
+    command: JsonAstNode<string> | undefined
+  ): Record<string, string> {
+    if (syntaxInfo.wireitConfigNode === undefined) {
+      return {};
+    }
+    const envNode = findNodeAtLocation(syntaxInfo.wireitConfigNode, ['env']);
+    if (envNode === undefined) {
+      return {};
+    }
+    if (command === undefined) {
```

Comment:
Does this mean passthrough commands can't set env vars?

What about a pair of `start` and `start:dev` scripts that are both pass-throughs where the only difference is that `start:dev` sets something like `MODE: "development"` for other scripts?

Reply:
> Does this mean passthrough commands can't set env vars?

Yep. I general `env` only applies to that specific script, it doesn't apply transitively, so it doesn't make sense if there is no `command`, since it would have no effect. (The non-transitiveness is documented in the README).

> What about a pair of `start` and `start:dev` scripts that are both pass-throughs where the only difference is that `start:dev` sets something like `MODE: "development"` for other scripts?

In that case I'd create two different branches of the dependencies. You'd probably want different output folders anyway.

Reply:
Hmm, maybe I should reopen https://github.com/google/wireit/issues/556

Because my question there is about how to avoid having a big duplication of scripts to handle the cross-cutting prod/dev cases. If I have separate scripts, then I don't have any need for an env variable to be an input to the fingerprint - I would either not use env vars, or each script would only ever be invoked with a single value of the variable.

Reply:
Another case where sharing env vars would be important is if we figure out how to start services with dynamic resources - like a server with a dynamic port. In that case we'd like to somehow extract the port and give it to anything that depends on it. That would allow for starting services multiple times (the opposite of what I'm asking for in https://github.com/google/wireit/issues/580, I think both behaviors would be useful).

This is two different directions of sharing: from dependent->dependency, and from dependency->dependent.

Taking just the dependent->dependency which I'm asking about originally, could we have a way to explicitly share the value via a property in dependencies?

```json
{
  "build" {
    "command": "...",
    "env": {
      "MODE": "external"
    }
    "files": [],
    "output": [],
  },
  "start": {
    "command": "...",
    "env": {
      "MODE": {
        "external": true,
    },
    "dependencies": {
      "build": {
        "env": {
          "MODE": {
            // pass "start"s MODE value to "build"
            "inherit": true,
          }
        }
      }
    }
  }
}
```

Reply:
Then any dependency of the dependency would also need the env option too if you need it to propagate. Would it be a bad idea to add something like "cascade" at the script's env config, where it could be like this?
```json
{
  "start": {
    "command": "...",
    "env": {
      "MODE": {
        "value": "prod",
        "cascade": true
      }
    }
  }
}
```

Reply:
If an environment variable applies to all dependencies, then that implies that that environment variable should be included in the fingerprint. But very often that's not desirable, because it would mean you can't re-use output even though the variable doesn't actually matter for some particular script.

For example, if you had a dev and prod rollup script which differ via an environment variable., but both of them consumed the same `tsc` output, then you'd lose the cache hit for the `tsc` step.

---

Things also get a bit confusing when you have multiple paths to the same dependency, e.g. imagine this graph:

```
   A
  / \
 v   v
 B   C [BUILD=prod]
  \ /
   v
   D
```

So now D as traversed via B has BUILD=undefined, but as traversed via C has BUILD=prod. It's not clear to me what we should do there, and I suspect the answer is we shouldn't get ourselves in that situation to begin with.

Reply:
Discussed some more offline with Justin -- I had actually missed the comment about setting `"inherit": true`, which addresses my concerns above pretty much. We also extended the idea to consider how you could directly reference the _specific_ script where you want to inherit an environment variable from. Also agreed that this can be a new extension to this feature, so we can land this initial support now.